File: Learning/Python/Python数据结构与算法/算法/排序/8快速排序/QuickSort2.py
#优化2: 优化不必要的交换

def QuickSort(k):
    QSort( k, 0, len(k)-1 )

def QSort(k, low, high):
    if ( low < high ):
        point = Partition(k, low, high)
        QSort(k, low, point-1)
        QSort(k, point+1, high)

def Partition(k, low, high):
    mid = low + (high-low)//2
    if k[low] > k[high]:
        swap(k, low, high)
    if k[mid] > k[high]:
        swap(k, mid, high)
    if k[low] < k[mid]:
        swap(k, low, mid)
        
    point = k[low]
    
    while( low < high ):
        while( k[high]>=point and low<high ):
            high -= 1
        k[low] = k[high]
        while( k[low]<=point and low<high ):
            low += 1
        k[high] = k[low]
    k[low] = point
    return low

def swap(k, high, low):
    k[low],k[high] = k[high],k[low]
    

k = [55, 2, 8, 1, 34, 21, 5, 3, 89, 13]
QuickSort(k)

print("从小到大排列为:")
for i in k:
    print(i, end=' ')

File: Learning/数据结构与算法/算法/排序/8快速排序/QuickSort2/QuickSort2.c
#include <stdio.h>

void swap(int k[], int low, int high)
{
    int temp;

    temp = k[low];
    k[low] = k[high];
    k[high] = temp;
}

int Partition(int k[], int low, int high)
{
    int point;

    int m = low + (high-low)/2;
    if(k[low] < k[high])
    {
        swap(k, low, high);
    }
    if(k[m] > k[high])
    {
        swap(k, m, high);
    }
    if(k[m] > k[low])
    {
        swap(k, m, low);
    }

    point = k[low];

    while( low < high )
    {
        while( low < high && k[high] >= point )
        {
            high--;
        }
        k[low] = k[high];
        while( low < high && k[low] <= point)
        {
            low++;
        }
        k[high] = k[low];
    }
    k[low] = point;

    return low;
}

void QSort(int k[], int low, int high)
{
    int point;

    if(low < high )
    {
        point = Partition(k, low, high);  //!关键:计算基准点

        QSort(k, low, point-1);

        QSort(k, point+1, high);
    }
}

void QuickSort(int k[], int n)
{
    QSort(k, 0, n-1);
}

int main()
{
    int i, a[]={5,2,6,0,3,9,1,7,4,8};
    int len=sizeof(a)/sizeof(int);
    QuickSort(a, len);

    printf("排序后的结果是:\n");
    for(i=0; i<len; i++)
    {
        printf("%d ", a[i]);
    }
    printf("\n\n");

}
